tests/agents/gherkin-specs/outcome.feature

```diff
@@ -0,0 +1,110 @@
+Feature: Outcome
+
+  # ---- user set outcome
+
+  Scenario: User set outcome on span has priority over instrumentation
+    Given an agent
+    And an active span
+    And user sets span outcome to 'failure'
+    And span terminates with outcome 'success'
+    Then span outcome is 'failure'
+
+  Scenario: User set outcome on transaction has priority over instrumentation
+    Given an agent
+    And an active transaction
+    And user sets transaction outcome to 'unknown'
+    And transaction terminates with outcome 'failure'
+    Then transaction outcome is 'unknown'
+
+  # ---- span & transaction outcome from reported errors
+
+  Scenario: span with error
+    Given an agent
+    And an active span
+    And span terminates with an error
+    Then span outcome is 'failure'
+
+  Scenario: span without error
+    Given an agent
+    And an active span
+    And span terminates without error
+    Then span outcome is 'success'
```

Comment:
This looks good to me for any type of span, but the [written spec](https://github.com/elastic/apm/blob/master/specs/agents/tracing-spans.md#span-outcome) allows `unknown` (or even `null`), with a note about the importance of exit spans. Does this depend on the span type?

Reply:
I think that's a good question here:

- the exit spans are the ones that call a third party in the map, for those we MUST have an outcome in 99% of cases
- we can provide a list of span types that are exit spans (`db`, `messaging`, ...)

However, adding the type here would mean to either duplicate those statements for each type or use a scenario outline (with examples) to test for those, and to me applying this rule also works with other types of spans:

- we might have non-exit spans that may trigger errors (like the ones captured through annotations in Java): for those reporting exceptions and mark the span as an error would also make sense, even if not an exit span.
- we have some spans (like inferred spans) which will never have a known outcome, and for those it might make sense to always have `unknown`.

Here we only test for the cases for which the outcome is known from the presence/absence of error, but it does not mean that we can't have unknown outcomes.